### DISCUSSION TITLE
Extend script with dynamic deployment mode

### DISCUSSION THREAD

**Discussion by Hardexit** [PSAppDeployToolkit/PSAppDeployToolkit#1191] [Q&A] [answered]
Hello,

in the new v4 there is the `PSAppDeployToolkit.Extensions`, now I would like to reintroduce my extension, see below, so that Intune can install the software without user interaction, but the user must interact if he is logged in.

Can I add this to the `PSAppDeployToolkit.Extensions` or do I still have to add it to the `Invoke-AppDeployToolkit.ps1`?

```powershell
    [String]$processName = "explorer"
    $runningProcesses = Get-Process | Where-Object { $_.ProcessName -eq $processName }
    if ($runningProcesses) {
        $regValue = Get-ItemProperty -Path $HKLMIntune -Name $appName -ErrorAction SilentlyContinue
        if ($null -ne $regValue) {
            [String]$DeployMode = 'Interactive'
        }
        else {
            [String]$DeployMode = 'NonInteractive'
        }
    }
    else {
        [String]$DeployMode = 'NonInteractive'
    }
```

**Comment by DanGough** [comment]
Have you tried this launcher script to launch the toolkit via ServiceUI if required under Intune?

https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/tree/main/examples/ServiceUI

**Reply by Hardexit** [reply]
Yes, that would be possible, but then I'm still missing the option of silent installation if the program has never been on the computer.

Currently I start it via intune like this: `ServiceUI.exe -Process:explorer.exe Invoke-AppDeployToolkit.exe -DeploymentType Install`

**Reply by mjr4077au** [reply]
@Hardexit I can't see your comment regarding "Der Wert kann nicht in den Typ "System.String" konvertiert werden.", but please use these examples as they contain crucial fixes for v4: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/tree/develop/examples/ServiceUI

**Reply by Hardexit** [reply]
that works, thank you very much.
This is not a comment, it was issued by ps in debug mode. `Set-PSDebug -Trace 2`

**Reply by Hardexit** [reply]
@mjr4077au i still see a problem with the helper script, it returns a pupup and not a pure value in the console, intune can't handle that and just runs into an unexpected error.

**Comment by mjr4077au** [comment]
> Hello,
> 
> in the new v4 there is the `PSAppDeployToolkit.Extensions`, now I would like to reintroduce my extension, see below, so that Intune can install the software without user interaction, but the user must interact if he is logged in.
> 
> Can I add this to the `PSAppDeployToolkit.Extensions` or do I still have to add it to the `Invoke-AppDeployToolkit.ps1`?
> 
> ```powershell
>     [String]$processName = "explorer"
>     $runningProcesses = Get-Process | Where-Object { $_.ProcessName -eq $processName }
>     if ($runningProcesses) {
>         $regValue = Get-ItemProperty -Path $HKLMIntune -Name $appName -ErrorAction SilentlyContinue
>         if ($null -ne $regValue) {
>             [String]$DeployMode = 'Interactive'
>         }
>         else {
>             [String]$DeployMode = 'NonInteractive'
>         }
>     }
>     else {
>         [String]$DeployMode = 'NonInteractive'
>     }
> ```

This type of code would have to fire before `$adtSession` is even defined to ensure the changed `$DeployMode` is captured properly. We could probably add some kind of option like this to `Open-ADTSession` where if you pass through process names, it fires off similar logic.

The code you provided could be simplified down to this as well:
```PowerShell
$DeployMode = if (Get-Process -Name explorer) {
    if (Get-ItemProperty -Path $HKLMIntune -Name $appName -ErrorAction Ignore) {
        'Interactive'
    }
    else {
        'NonInteractive'
    }
}
else {
    'NonInteractive'
}
```
Doing `Get-Process -Name` instead of `Get-Process | Where-Object {}` is faster as you're filtering in the command, not retrieving all results and ditching afterwards.

Using `-ErrorAction Ignore` is preferable to `SilentlyContinue` the latter still writes the error to PowerShell's `$Error` array, which has a cost. Using `Ignore` bypasses all of this, which is what you want anyway.

Lastly, it's little-known, but in PowerShell the output of an if statement can be saved into a variable. This means you don't need to define `[String]$DeployMode = ` three times, you can just do it on the entire if statement itself.

Not trying to tell you what's what or anything, just some tips in case you didn't know.

**Reply by Hardexit** [reply]
Hi, thanks for the tips, I'll give it a try. What is the difference between NonInteractive and Silent? Because I was wondering if Silent would be the better option for when no explorer is active.

**Reply by mjr4077au** [reply]
In 3.x, basically nothing. However in v4, NonInteractive acts like a passive setup, where you can see progress but there's no prompts or questions asked.

**Reply by Hardexit** [reply]
Thank you very much, unfortunately I did not find this information. Then it would make sense now:

```
$DeployMode = if (Get-Process -Name explorer) {
    if (Get-ItemProperty -Path $HKLMIntune -Name $appName -ErrorAction Ignore) {
        'Interactive'
    }
    else {
        'NonInteractive'
    }
}
else {
    'Silent'
}
```

**Reply by mjr4077au** [reply]
If you don't want people to see anything at all (and if there's no process running, there's no real need for them to see anything anyway), you can just set the other `NonInteractive` to `Silent` if you wish.

**Reply by Hardexit** [reply]
At that moment it is not possible to see anything because everything is covered by an Intune overlay. I don't know if you know this, but it looks like this:
![Image-min](https://github.com/user-attachments/assets/9c9daabd-f3a8-4149-99e7-12883e25547c)

**Reply by mjr4077au** [reply]
I believe `0x87d1041c` means that the application wasn't detected after a successful installation.

**Reply by Hardexit** [reply]
No not quite, the problem was that I did not customize the installation command in Intune and the file has been renamed.

**Reply by Hardexit** [reply]
I just realized that my snippet doesn't work anymore because I can't read the AppName and write the DeployMode because both are in the $adtSession.

This has reduced the whole thing a little 😅
```
$DeployMode = if (-not (Get-Process -Name explorer -ErrorAction Ignore)) {
    'Silent'
}
```

**Reply by JonathanPitre** [reply]
@Hardexit can you please let me know where you would add that code snippet exactly? thanks!

**Reply by Hardexit** [reply]
Hi, I have completely discarded it and used the [script](https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/tree/develop/examples/ServiceUI) here.

Unfortunately, the explorer.exe is already present when the following screen is displayed:
![Image-min](https://github.com/user-attachments/assets/9c9daabd-f3a8-4149-99e7-12883e25547c)

So the only possibility I see is to fall back on the registration keys I set, but unfortunately I have the problem that I have to put the code before the `$adtSession` but need values from the `$adtSession`.

You could be a workaround:
```
[String]$appVendor = 'Vendor'
[String]$appName = 'Name'
[String]$appVersion = 'v1.0.0'
[String]$appFile = 'FILE'
[String]$appParameters = 'VERYSILENT' # Ohne Parameter Leerzeichen
[String]$CloseApps = 'PROZESSNAME'
[String]$uuid = ''

[String]$regkey = 'machine' # machine or user

if ($regkey -eq 'machine') {
    [String]$RegKeyPath = 'HKLM:\SOFTWARE\Microsoft\Intune'
} else {
    [String]$RegKeyPath = 'HKCU:\SOFTWARE\Microsoft\Intune'
}

$DeployMode = if (Get-ItemProperty -Path $RegKeyPath -Name $appName -ErrorAction Ignore) {
    'Interactive'
}
else {
    'NonInteractive'
}

##================================================
## MARK: Variables
##================================================

$adtSession = @{
    # App variables.
    AppVendor = $appVendor
    AppName = $appName
    AppVersion = $appVersion
```

**Reply by Hardexit** [reply]
I've come up with a new idea on how to handle it, using this method. However, this requires 4.0.4, as I have added variables to the `$adtSession`.

```
$adtSession = @{
    # App variables.
    AppVendor = 'VENDOR'
    AppName = 'NAME'
    AppVersion = 'VERSION'
    appFile = 'APPFILE'
    appParameters = 'PARAMETER' # Ohne Parameter Leerzeichen
    CloseApps = 'CLOSEAPP'
    AppArch = ''
    AppLang = ''
    AppRevision = '01'
    AppSuccessExitCodes = @(0)
    AppRebootExitCodes = @(1641, 3010)
    AppScriptVersion = '1.0.0'
    AppScriptDate = '2024-12-09'
    AppScriptAuthor = 'AppScriptAuthor'

    uuid = ''

    regkey = 'machine' # machine or user

    wingetID = 'false' # false or wigetID

    # Install Titles (Only set here to override defaults set by the toolkit).
    InstallName = ''
    InstallTitle = ''

    # Script variables.
    DeployAppScriptFriendlyName = $MyInvocation.MyCommand.Name
    DeployAppScriptVersion = '4.0.4'
    DeployAppScriptParameters = $PSBoundParameters
}

if ($($adtSession.regkey) -eq 'machine') {
    [String]$RegKeyPath = 'HKLM:\SOFTWARE\Microsoft\Intune'
} else {
    [String]$RegKeyPath = 'HKCU:\SOFTWARE\Microsoft\Intune'
}

if (Get-ItemProperty -Path $RegKeyPath -Name $($adtSession.appName) -ErrorAction Ignore) {
    $adtSession["DeployMode"] = "Interactive"
}
else {
    $adtSession["DeployMode"] = "NonInteractive"
}
```

**Reply by mjr4077au** [reply]
@Hardexit I can see you've since edited the post quite a few times since the initial message, however the reason the variables were removed from `$adtSession` is that they no longer need to be there with the changes made in https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/380dfc9e2e8c9c28cd060fceaeeb9b41af34fc4b.

**Comment by mjr4077au** [comment, marked as answer]
The request in the initial question is being tracked here: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/issues/1297. Closing this discussion for now.
